### ISSUE TITLE
Cargo suggestion doesn't work with crates having a hyphen in the name

### ISSUE THREAD

**Issue by hrxi** [rust-lang/rust#115841]
While hitting https://github.com/rust-lang/rust/issues/84970#issuecomment-1719441857, I got this from Cargo:
```
$ cargo check
[…]
error: internal compiler error: encountered incremental compilation error with mir_built(2b0ee002860f4e6d-86dbf79d6866cd33)
  |
  = help: This is a known issue with the compiler. Run `cargo clean -p nimiq_lib` or `cargo clean` to allow your project to compile
  = note: Please follow the instructions below to create a bug report with the provided information
  = note: See <https://github.com/rust-lang/rust/issues/84970> for more information
[…]
```

Running `cargo clean -p nimiq_lib` failed with the following error:
```
$ cargo clean -p nimiq_lib
error: package ID specification `nimiq_lib` did not match any packages

	Did you mean `nimiq-lib`?
$ cargo clean -p nimiq-lib
$
```
And indeed `cargo clean -p nimiq-lib` works.

**Comment by apiraino** [comment]
Not completely sure but https://github.com/rust-lang/cargo/issues/9038 could be related. Probably this report can be moved to that repository.

**Comment by weihanglo** [comment]
The error message is from rustc diagnostics:

https://github.com/rust-lang/rust/blob/376f3f059059b7fdd1331c77103943fd7c22918c/compiler/rustc_query_system/src/query/plumbing.rs#L700-L710

And `--crate-name` is provided with hyphen-to-underscore normalization to rustc by cargo.

I've got the impression that the Cargo team tries to reduce such kind of normalization in exchange for performance, at least for crates.io index query. Package ID flag `-p/--package` is also a thing that never got normalized. Thus, I lean toward not fixing this to avoid complexity.

Given the above points, and the helpful `Did you mean …?` message, I propose to close this as "wont fix" at this moment.

**Comment by hrxi** [comment]
So the problem is that rustc tries to guess the correct cargo invocation based on incomplete (hyphen-to-underscore-normalized) information passed to it? Cargo doesn't want to do reverse-normalization on `-p`, so the only option would be to pass the non-normalized data to rustc, but that's considered too complex?

**Comment by weihanglo** [comment]
> so the only option would be to pass the non-normalized data to rustc, but that's considered too complex?

This won't be too complex I guess. It just seems too specialized to me. We need to add a new rustc option and make cargo pass it only for this incremental compilation error. It also kinda leaks the abstraction between Cargo and rustc. Not sure if that is a right approach for this.

**Comment by Enselic** [comment]
> I propose to close this as "wont fix" at this moment.

No objections were raised, so let's close.

**Comment by hrxi** [comment]
Is it not useful to keep it open as the problem itself still exists, even though we don't want to fix it right now?

**Comment by Enselic** [comment]
If you disagree with closing I'm fine with reopening.
